src/goto-programs/goto_convert.cpp

```diff
@@ -93,6 +93,141 @@ static void finish_catch_push_targets(goto_programt &dest)
   }
 }
 
+struct build_declaration_hops_inputst
+{
+  irep_idt mode;
+  irep_idt label;
+  goto_programt::targett goto_instruction;
+  goto_programt::targett label_instruction;
+  node_indext label_scope_index = 0;
+  node_indext end_scope_index = 0;
+};
+
+void goto_convertt::build_declaration_hops(
+  goto_programt &program,
+  std::unordered_map<irep_idt, symbolt, irep_id_hash> &label_flags,
+  const build_declaration_hops_inputst &inputs)
+{
+  // In the case of a goto jumping into a scope, the declarations (but not the
+  // initialisations) need to be executed. This function performs a
+  // transformation from code that looks like -
+  //   {
+  //     statement_block_a();
+  //     if(...)
+  //       goto user_label;
+  //     statement_block_b();
+  //     int x;
+  //     x = 0;
+  //     statement_block_c();
+  //     int y;
+  //     y = 0;
+  //     statement_block_d();
+  //   user_label:
+  //     statement_block_e();
+  //   }
+  // to code which looks like -
+  //   {
+  //     __CPROVER_bool going_to::user_label;
```

Comment:
I am not convinced that something that can be solved at compile time should introduce runtime cost.

Reply:
I feel confident that the change proposed in this PR will result in an analysis which is correct. I am not certain that the the issue can be solved at compile time in all cases. As such I would propose that we fix first and that optimisation should be considered as a follow-up if there is indeed a notable runtime cost. Please also bear in mind that this transformation is only performed in the case of a goto jumping into a scope and not for other constructs.

If you look at the `regression/cbmc/declaration-goto/re-enter_scope.c` included in this PR then I would say that -

1. The `assert(i == 42)` assertion should fail. The scope of the `i` and `j` variables does not intersect, therefore a C compiler should be free to use the same stack location for both variables. Therefore the `i` variable is not guaranteed to hold the same value on re-entering the scope.
2. If we attempt to solve this problem in cbmc by moving the declaration of the `i` variable alone, then cbmc would report `SUCCESS` for `assert(i == 42)` when it should fail.